src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,133 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private T[] arrayList;
+    int size = 0;
+    int emptyPlaces = 10;
+
+    public ArrayList() {
+        arrayList = (T[]) new Object[emptyPlaces];
+    }
+
+    public void enlarge() {
+        emptyPlaces = (int) (size * 0.5);
+        T[] temporaryEnlargedArrayList = (T[]) new Object[size + emptyPlaces];
+        for (int i = 0; i < size; i++) {
+            temporaryEnlargedArrayList[i] = arrayList[i];
+        }
+        arrayList = (T[]) new Object[size + emptyPlaces];
+        for (int i = 0; i < size; i++) {
+            arrayList[i] = temporaryEnlargedArrayList[i];
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (emptyPlaces == 0) {
+            enlarge();
+        }
+        arrayList[size] = value;
+        emptyPlaces--;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("given index does not exist");
+        } else if (index == size - 1) {
+            set(value, size -1);
+        } else {
+            T[] temporaryArrayList = (T[]) new Object[size - index];
+            for (int i = index; i < size; i++) {
+                temporaryArrayList[i - index] = arrayList[i];
+            }
+            set(value, index);
+            for (int i = size - 1; i > index; i--) {
+                remove(i);
+            }
+            for (T element : temporaryArrayList) {
+                add(element);
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
 
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("given index does not exist");
+        } else {
+            return arrayList[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("given index does not exist");
+        } else {
+            arrayList[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
```

Comment:
The index check logic is repeated in multiple methods. Extract this into a separate method to avoid code duplication.